File: ruby/82_remove_duplicates_from_sorted_list_two.rb
# Definition for singly-linked list.
class ListNode
    attr_accessor :val, :next
    def initialize(val, nextNode=nil)
        @val = val
        @next = nextNode
    end
end


# @param {ListNode} head
# @return {ListNode}

# Input: 1->2->3->3->4->4->5
# Output: 1->2->5
# Input: 1->1->1->2->3
# Output: 2->3

# 给定链表，删除重复元素，不留一个

def delete_duplicates(head)
    return head if head == nil || head.next == nil

    if head.val == head.next.val
        num = head.val 
        while head && head.val == num
            head = head.next
        end 
        return delete_duplicates(head)
    else
        head.next = delete_duplicates(head.next)
        return head
    end
end

def delete_duplicates_version2(head)
    return head if head == nil || head.next == nil

    p, q = nil, head
    while q && q.next
        if q.val == q.next.val
            p = q

            while q.next && q.val == q.next.val
                q = q.next
            end

            if p == head
                head = q.next
            else
                node = head
                while node.next != p
                    node = node.next
                end 
                node.next = q.next
            end 
        end

        q = q.next
    end

    head
end 

def show_list_node_val(node)
	str = ""
	while node
		str += node.val.to_s
		node = node.next
		str += '->' if node
	end

	str
end

l1 = ListNode.new(1, ListNode.new(1, ListNode.new(2, nil)))

l2 = delete_duplicates(l1)

puts show_list_node_val(l2)

File: cpp/82_remove_duplicates_from_sorted_list_two.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// 给定链表，要求删除所有出现超过一次的节点；
// 1> 使用map记录次数，然后构建新链表；
// 2> 链表移动，构建三个节点，第一个节点删除前，第二个是遍历的节点，第三个是重复的最后一个节点；
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
		if (!head || !head->next) { return head; }

		ListNode *tmp = new ListNode(head->val-1);
		tmp->next = head;

		ListNode *first = tmp;
		ListNode *second = tmp->next;
		ListNode *third = NULL;

		int val = 0;
		while (second) {
			val = second->val;
			auto nextNode = second->next;
			if (!nextNode) { break; }

			if (nextNode->val == val) {
				third = nextNode;
			} else {
				if (third) {
					first->next = third->next;
					third = NULL;
				} else {
					first = second;
				}
			}
			second = nextNode;
		}
		if (third) { first->next = third->next; }
		return tmp->next;
    }
};
